pkg/controller/wildflyserver/wildflyserver_controller.go

```diff
@@ -130,12 +170,161 @@ func (r *ReconcileWildFlyServer) Reconcile(request reconcile.Request) (reconcile
+	}
+
+	// Checking if  WildFlyServerSpec is about to be scaled down
+	podList, err := r.getPodsForWildFly(wildflyServer)
+	if err != nil {
+		reqLogger.Error(err, "Failed to list pods.", "WildFlyServer.Namespace", wildflyServer.Namespace, "WildFlyServer.Name", wildflyServer.Name)
 		return reconcile.Result{}, err
 	}
+	wildflyServerSpecSize := wildflyServer.Spec.Size
+	numberOfPodsToScaleDown := len(podList.Items) - int(wildflyServerSpecSize) // difference between desired pod count and the current number of pod instances
+	scaleDownPodsState := make(map[string]string)                              // map referring to: pod name - pod state
+	scaleDownErrors := []error{}                                               // errors occured during processing the scaledown for the pods
+	for scaleDownIndex := 1; scaleDownIndex <= numberOfPodsToScaleDown; scaleDownIndex++ {
+		// Scaledown scenario, need to handle transction recovery
+		scaleDownPod := podList.Items[len(podList.Items)-scaleDownIndex]
+		scaleDownPodName := scaleDownPod.ObjectMeta.Name
+		scaleDownPodIP := scaleDownPod.Status.PodIP
+		if strings.Contains(scaleDownPodIP, ":") && !strings.HasPrefix(scaleDownPodIP, "[") {
+			scaleDownPodIP = "[" + scaleDownPodIP + "]" // for IPv6
+		}
+
+		// Setting-up the pod status - status is used to decide if the pod could be scaled (aka. removed from the statefulset)
+		wildflyServerSpecPodStatus := getWildflyServerPodStatusByName(wildflyServer, scaleDownPodName)
+		scaleDownPodsState[scaleDownPodName] = wildflyServerSpecPodStatus.State
+		if wildflyServerSpecPodStatus.State != wildflyv1alpha1.PodStateScalingDownClean {
+			// When state is not already processed with recovery and all data was cleaned then we mark pod to be dirty
+			scaleDownPodsState[scaleDownPodName] = wildflyv1alpha1.PodStateScalingDownDirty
+		}
+
+		if wildflyServerSpecPodStatus.State != wildflyv1alpha1.PodStateScalingDownClean {
+			// Removing the pod from the Service handling
+			scaleDownPod.ObjectMeta.Labels[labelWildflyOperatorInService] = "under-scale-down-processing"
+			if err = r.client.Update(context.TODO(), &scaleDownPod); err != nil {
+				reqLogger.Error(err, "Failed to update pod labels", "Pod name", scaleDownPod.ObjectMeta.Name, "Pod labels ", scaleDownPod.ObjectMeta.Labels)
+				scaleDownErrors = append(scaleDownErrors, err)
+				break
+			}
+
+			reqLogger.Info("Going for scaledown with pod", "pod name", scaleDownPod.ObjectMeta.Name, "IP address", scaleDownPodIP)
+
+			managementURL := fmt.Sprintf("http://%v:%v/management", scaleDownPodIP, httpManagementPort)
+			username := string(foundTxnRecoverySecret.Data["username"])
+			password := string(foundTxnRecoverySecret.Data["password"])
+			httpDigest := wildflyutil.HTTPDigest{URL: managementURL, Username: username, Password: password}
+
+			// Enabling recovery listener to speed up recovery processing
+			jsonResult, err := wildflyutil.ExecuteMgmtOp(&httpDigest, wildflyutil.MgmtOpTxnEnableRecoveryListener)
```

Comment:
could you make a CLI exec instead?

Reply:
I could but I thought the remote management operation decouples the communication better from the underlaying container. If you think it's better/nicer/whatever I can change all the calls to use the `CLI exec` instead. Just please let me know what is your preference and I can do it.